### DISCUSSION TITLE
Placeholders in base URL

### DISCUSSION THREAD

**Discussion by BeyondEvil** [encode/httpx#2679] [Q&A] [answered]
Is it possible to provide a base URL to Client with placeholders (similar to f-strings/format)? Either named or positional?

Something like: `Client(base_url="https://dev.azure.com/{org_id}/{project_id}/_apis/wit/workitems/")`

**Comment by lovelydinosaur** [comment]
I thought it'd be interesting to see what ChatGPT says...

<img width="867" alt="Screenshot 2023-04-26 at 13 56 33" src="https://user-images.githubusercontent.com/647359/234581614-5aac1534-d0c4-43e0-8d90-71d8bd6c6346.png">

<img width="855" alt="Screenshot 2023-04-26 at 13 59 44" src="https://user-images.githubusercontent.com/647359/234582505-d5658451-0c07-4a21-aaee-7a88481fa74a.png">

<img width="765" alt="Screenshot 2023-04-26 at 13 59 50" src="https://user-images.githubusercontent.com/647359/234582561-782cc962-0cf0-4bc6-8648-e702b0ee9dfb.png">

**Comment by BeyondEvil** [comment, marked as answer]
Yeah, that's similar to what I did, but I have a function.

```python
from httpx import Client

from syncer import config


class HTTPClient:
    def __init__(self):
        self._client = None
        self._default_headers = {"Content-Type": "application/json"}
        self._default_params = {"api-version": "7.0"}

    def client(self):
        if self._client is None:
            self._client = Client(
                auth=("", config.get("pat")),
                headers=self._default_headers,
                params=self._default_params,
            )
        return self._client

def url(org_id, project_id, endpoint):
    return f"https://dev.azure.com/{org_id}/{project_id}/_apis/wit/workitems/{endpoint}"
```

and a call might look like:

```python
response = self._http_client.delete(
            url(self.org, self.project_id, self.id),
        )
```

**Reply by BeyondEvil** [reply]
I guess I could remove the `endpoint` as, IIRC, I can pass that in as an arg to the different verbs.

**Comment by BeyondEvil** [comment]
Would you review a PR providing this feature @tomchristie ?
